drivers/backup/controllers/cluster/cluster_controller.go

```diff
@@ -0,0 +1,255 @@
+		return &NamespaceConfig{
+			namespace: namespace,
+		}
+	}
+	namespaceInfo := c.getNamespaceInfo(namespace)
+	return &NamespaceConfig{
+		namespace:           namespace,
+		isNamespaceRecorded: true,
+		contexts:            namespaceInfo.contexts,
+		DestroySchedulerContextConfig: &DestroySchedulerContextConfig{
+			waitForDestroy:             DefaultWaitForDestroy,
+			waitForResourceLeakCleanup: DefaultWaitForResourceLeakCleanup,
+			skipClusterScopedObjects:   DefaultSkipClusterScopedObjects,
+		},
+		ValidateSchedulerContextConfig: &ValidateSchedulerContextConfig{
+			waitForRunningTimeout:       DefaultWaitForRunningTimeout,
+			waitForRunningRetryInterval: DefaultWaitForRunningRetryInterval,
+			validateVolumeTimeout:       DefaultValidateVolumeTimeout,
+			validateVolumeRetryInterval: DefaultValidateVolumeRetryInterval,
+		},
+		controller: c,
+	}
+}
+
+// Cleanup cleans up all the cluster resources recorded in the ClusterController
+func (c *ClusterController) Cleanup() (err error) {
+	defer func() {
+		err = utils.SwitchClusterContext(GlobalInClusterConfigPath)
+	}()
+	namespaces := c.getAllNamespaces()
+	for _, namespace := range namespaces {
+		log.Infof("Cleaning up namespace [%s]", namespace)
+		err = c.Namespace(namespace).Destroy()
+		if err != nil {
+			debugMessage := fmt.Sprintf("namespace: [%s]", namespace)
+			return utils.ProcessError(err, debugMessage)
+		}
+	}
+	c.appKeyCountMap = make(map[string]int, 0)
+	return nil
+}
+
+// Cluster returns a new ClusterInfo instance with the specified cluster id, name, and config-path
+func Cluster(id int, name string, configPath string) *ClusterInfo {
+	return &ClusterInfo{
+		ClusterMetaData: &ClusterMetaData{
+			Id:         fmt.Sprintf("%d", id),
+			Name:       name,
+			ConfigPath: configPath,
+		},
+	}
+}
+
+// AddClusterControllersToMap adds ClusterController instances to the specified map based on ClusterInfo objects, using `ClusterInfo.Name` as the key
+func AddClusterControllersToMap(clusterControllerMap *map[string]*ClusterController, clustersInfo []*ClusterInfo) error {
+	if *clusterControllerMap == nil {
+		*clusterControllerMap = make(map[string]*ClusterController)
+	}
+	inClusterInfo := Cluster(0, utils.DefaultInClusterName, GlobalInClusterConfigPath).IsHyperConverged()
+	clustersInfo = append(clustersInfo, inClusterInfo)
+	for _, clusterInfo := range clustersInfo {
+		clusterController, err := clusterInfo.GetController()
+		if err != nil {
+			debugMessage := fmt.Sprintf("cluster: name [%s], config path [%s]", clusterInfo.Name, clusterInfo.ConfigPath)
+			return utils.ProcessError(err, debugMessage)
+		}
+		(*clusterControllerMap)[clusterInfo.Name] = clusterController
+	}
+	return nil
+}
+
+// AddTestCaseClusterControllers adds ClusterController instances associated with the specified testRailId for use by a test case
+func AddTestCaseClusterControllers(clusterControllerMap *map[string]*ClusterController, testRailId int) error {
```

Comment:
Can we rename this to `SetClusterController`?

Reply:
Also rename `testRailId` to just `id` or `uniqueId`

Reply:
Done. Changed AddTestCaseClusterControllers to AddDefaultClusterControllers.